src/Tests/library/test_rwlocks.pl

```diff
@@ -0,0 +1,112 @@
+       notice, this list of conditions and the following disclaimer in
+       the documentation and/or other materials provided with the
+       distribution.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+
+:- asserta(user:file_search_path(library, './library')).
+:- use_module(library(rwlocks)).
+
+:- begin_tests(rwlocks).                                                                                                                                          
+                                                                                                                                                                      
+    test(priority_handling) :-
+        retractall(started(_)),
+        with_rwlock(test_lock, 
+                    (assertz(started(write)), 
+                    sleep(1), 
+                    writeln('Write operation completed')), 
+                    write),
+        with_rwlock(test_lock, 
+                    (assertz(started(read)), 
+                    sleep(2), 
+                    writeln('Read operation completed')), 
+                    read),
+        findall(X, started(X), Started),
+        writeln(Started),
+        assertion(Started == [write, read]).                                                                                                                    
+                                                                                                                                                                      
+    test(timeout_handling, [throws(time_limit_exceeded(rwlock))]) :-
+        thread_create(with_rwlock(test_lock, sleep(2), write), WriteThread, []),
+        sleep(0.1),  % Give some time for the write lock to be acquired
+        catch(with_rwlock(test_lock, true, write, [timeout(0.5)]),
+            time_limit_exceeded(rwlock),
+            ( thread_signal(WriteThread, throw(abort_timeout_test)),
+                retract_rwlock_data(test_lock),
+                rethrow(time_limit_exceeded(rwlock))
+            )),
+        thread_join(WriteThread, _).
+
+    % Helper predicate to clean up any leftover rwlock data
+    retract_rwlock_data(LockId) :-
+        retractall(access(LockId, _, _)),
```

Comment:
More in general, all dynamic data should be declared using dynamic/1, and if possible, using thread_local/1 to improve thread-safety.   Ideally (not always possible), we'd like tests to be thread-safe, so we can run them concurrently.    This is faster and does a lot of additional testing for free.